### DISCUSSION TITLE
What's the expected behavior when using shared subscriptions in a single docker container?

### DISCUSSION THREAD

**Discussion by matthewadams** [vernemq/vernemq#2273] [Q&A] [answered]
We have integration tests that run against a single local docker container running vernemq (that is configured with a custom lua script for auth script, but I don't think this is involved).  We want to leave the shared subscription configuration the same when running in the test configuration as when we're running in a cluster.  That is, we don't want to use `foo/bar/baz/#` for a topic filter during integration testing, and `$shared/snafu/foo/bar/baz/#` when running in a cluster.  The vernmq config option `shared_subscription_policy` is set to `local_only` when we run the single container, `prefer_local` in the cluster.

When we run with a subscriber subscribing to a shared subscription, the subscriber never gets any messages when running vernemq as a single local container.  What's the expected behavior?  I expected it to just work, but it doesn't appear to be.  Our custom auth logs show the broker happily accepting the shared subscription without warning:

```
18:40:08.319 [debug] [custom-auth]     {"context":{"attempted_subscription_topics":["$shared/observation-event-mqtt-message-processor-adapter/canonical/json/observation/#"],"client_id":"site-app[gd45qkqqek2y499yyvhhdv7cfmk]"},"message":"🔗 Incoming subscription request.","timestamp":"2024-04-01 18:40:08"}
18:40:08.326 [info] [custom-auth]     {"context":{"attempted_subscription_topics":["$shared/observation-event-mqtt-message-processor-adapter/canonical/json/observation/#"],"client_id":"site-app[gd45qkqqek2y499yyvhhdv7cfmk]","subscribable_topic_filters":["#"]},"message":"✅  Allowing client topic subscriptions.","timestamp":"2024-04-01 18:40:08"}
```
There are many duplicated log messages like the following:
```
18:40:10.373 [debug] Replica meta5: Can't initialize AE exchange due to no peer available
```
but that is expected, I think, because there's only one broker in this configuration.

Thanks in advance.

**Comment by matthewadams** [comment, marked as answer]
Never mind.  Typo in shared subject name.  I had `$shared` (with a `d`) instead of `$share`.  :/

**Reply by ioolkos** [reply]
@matthewadams Thanks for being faster than me :)

Btw, yes the `18:40:10.373 [debug] Replica meta5: Can't initialize AE exchange due to no peer available` line is expected in a single broker (running on debug log). We could probably optimise this away on a single node, as the exchange FSMs could stay entirely inactive.
